SQLite 시작(대문자는 SQL문, 대문자+소문자는 데이터베이스명, 소문자는 필드명, 조건명

In [67]:
import sqlite3

In [68]:
conn = sqlite3.connect('./test.db')

SQLite의 이글스 테이블 지우기

In [8]:
cur = conn.cursor() # 커서 정의하기

In [9]:
cur.execute('drop table eagles')

데이터베이스 테이블 정의하기

In [17]:
cur.execute('CREATE TABLE IF NOT EXISTS Eagles \
    (back_no INT NOT NULL, \
     name TEXT, \
     position TEXT, \
     hands TEXT, \
     highschool TEXT, \
     height INT, \
     PRIMARY KEY(back_no));')

모든 필드 값이 있는 행을 테이블에 집어넣기

In [18]:
cur = conn.cursor()
cur.execute("INSERT INTO Eagles VALUES \
    (1, '하주석', '내야수', '우투좌타', '신일고', 184), \
    (28, '양성우', '외야수', '우투좌타', '충암고', 177);")

변경사항 저장하기

In [19]:
conn.commit()

파이썬에 표 집어넣기

In [21]:
import pandas as pd
players = pd.read_csv('./players.csv', encoding='EUC-KR')
players

back_no   name  position  hands  hightschool   height
0       13    최재훈        포수   우투우타          덕수고      178
1       30  제라드호잉       외야수   우투좌타       톨레도대학교      190
2       50    이성열       내야수   우투좌타          효천고      185
3       52    김태균       내야수   우투우타        천안북일고      185
4       57    정우람        투수   좌투좌타         경남상고      181
5       43    정은원       내야수   우투좌타          인천고      177
6       27    변우혁       내야수   우투우타          북일고      185
7       37    노시환       내야수   우투우타          경남고      185
8       66    유장혁       외야수   우투우타         광주일고      186
9       14    강경학       내야수   우투좌타        광주동성고      181

SQLite에 표 집어넣기

In [23]:
sql = 'INSERT INTO Eagles VALUES(?, ?, ?, ?, ?, ?);'
for i in range(10):
    cur.execute(sql, (int(players.iloc[i,0]),
                      players.iloc[i,1], players.iloc[i,2],
                      players.iloc[i,3], players.iloc[i,4],
                      int(players.iloc[i,5])))

In [24]:
conn.commit()

순회조회

In [26]:
cur.execute('SELECT * FROM Eagles')
for row in cur:
    print(row)

(1, '하주석', '내야수', '우투좌타', '신일고', 184)
(28, '양성우', '외야수', '우투좌타', '충암고', 177)
(13, '최재훈', '포수', '우투우타', '덕수고', 178)
(30, '제라드호잉', '외야수', '우투좌타', '톨레도대학교', 190)
(50, '이성열', '내야수', '우투좌타', '효천고', 185)
(52, '김태균', '내야수', '우투우타', '천안북일고', 185)
(57, '정우람', '투수', '좌투좌타', '경남상고', 181)
(43, '정은원', '내야수', '우투좌타', '인천고', 177)
(27, '변우혁', '내야수', '우투우타', '북일고', 185)
(37, '노시환', '내야수', '우투우타', '경남고', 185)
(66, '유장혁', '외야수', '우투우타', '광주일고', 186)
(14, '강경학', '내야수', '우투좌타', '광주동성고', 181)


데이터 조회는 fetch로 하면 됨 fetch(one, many, all)

In [29]:
cur.execute('SELECT * FROM Eagles')
row = cur.fetchone()
print(row)

(1, '하주석', '내야수', '우투좌타', '신일고', 184)


In [30]:
rows = cur.fetchmany(2)
for row in rows:
    print(row)

(28, '양성우', '외야수', '우투좌타', '충암고', 177)
(13, '최재훈', '포수', '우투우타', '덕수고', 178)


In [31]:
rows = cur.fetchall()
for row in rows:
    print(row)

(30, '제라드호잉', '외야수', '우투좌타', '톨레도대학교', 190)
(50, '이성열', '내야수', '우투좌타', '효천고', 185)
(52, '김태균', '내야수', '우투우타', '천안북일고', 185)
(57, '정우람', '투수', '좌투좌타', '경남상고', 181)
(43, '정은원', '내야수', '우투좌타', '인천고', 177)
(27, '변우혁', '내야수', '우투우타', '북일고', 185)
(37, '노시환', '내야수', '우투우타', '경남고', 185)
(66, '유장혁', '외야수', '우투우타', '광주일고', 186)
(14, '강경학', '내야수', '우투좌타', '광주동성고', 181)


In [32]:
cur.execute("SELECT name FROM Eagles WHERE height >180")
rows = cur.fetchall();
for row in rows:
    print(row)

('하주석',)
('제라드호잉',)
('이성열',)
('김태균',)
('정우람',)
('변우혁',)
('노시환',)
('유장혁',)
('강경학',)


In [35]:
cur.execute('SELECT * FROM Eagles ORDER BY name')
row = cur.fetchone()
print(row)

(14, '강경학', '내야수', '우투좌타', '광주동성고', 181)


In [36]:
cur.execute('SELECT * FROM Eagles ORDER BY name DESC')
row = cur.fetchone()
print(row)

(1, '하주석', '내야수', '우투좌타', '신일고', 184)


In [58]:
cur = conn.cursor()
cur.execute('SELECT * FROM Eagles ORDER BY height DESC LIMIT 5')
row = cur.fetchall();
for row in rows:
    print(row)

('내야수', 7, 183.14285714285714)
('외야수', 3, 184.33333333333334)
('투수', 1, 181.0)
('포수', 1, 178.0)


In [59]:
cur.execute('SELECT count(*) FROM Eagles')
count = cur.fetchone()
print(count)

(12,)


그루핑, 집계함수

In [60]:
cur.execute('SELECT position, count(*), avg(height) FROM Eagles GROUP BY position')
rows = cur.fetchall();
for row in rows:
    print(row)

('내야수', 7, 183.14285714285714)
('외야수', 3, 184.33333333333334)
('투수', 1, 181.0)
('포수', 1, 178.0)


placeholder 이용하기

In [62]:
cur = conn.cursor()
back_no = 50
cur.execute('SELECT * FROM Eagles WHERE back_no= ?;', (back_no,))
player = cur.fetchone()
print(player)

(50, '이성열', '내야수', '우투좌타', '효천고', 185)


업데이트 할 데이터가 없어서 마크다운으로 작성

=> 등번호 30번인 사람을 hands를 우투좌타로, hightschool은 미쿡고로 height는 190으로
수정하여라, 그리고 업데이트하여라

cur = coonn.cursonr()

cur.execute("UPDATE Eagles SET hands = '우투좌타', highschool='미쿡고', height=190 WHERE back_no = 30:")

conn.commit()


In [64]:
cur = conn.cursor()
cur.execute('DELETE FROM Eagles WHERE back_no=14;')
conn.commit()

table join을 위한 준비작업

In [91]:
cur = conn.cursor()
cur.execute('CREATE TABLE IF NOT EXISTS Stats \
    (id INT NOT NULL, \
     player TEXT, \
     average REAL, \
     rbi INT, \
     homerun INT, \
     PRIMARY KEY(id));')

In [92]:
conn.commit()

In [81]:
stats = pd.read_csv('./stats.csv', encoding='EUC-KR')
stats

player  average  rbi  homerun
0    최재훈    0.300   23        2
1  제라드호잉    0.288   58       15
2    이성열    0.255   62       18
3    김태균    0.308   43        5
4    정은원    0.281   47        5
5    변우혁    0.226    2        1
6    노시환    0.190   12        1
7    유장혁    0.174    3        1
8    강경학    0.213    6        0

In [93]:
cur = conn.cursor()
sql = 'INSERT INTO STATS VALUES (?, ?, ?, ?, ?);'
for i in range(9):
    cur.execute(sql, (i+1, 
                      stats.iloc[i,0], 
                      float(stats.iloc[i,1]),
                     int(stats.iloc[i,2]),
                     int(stats.iloc[i,3])))
conn.commit()

In [94]:
cur = conn.cursor()
cur.execute('SELECT * FROM Stats')
for row in cur:
    print(row)

(1, '최재훈', 0.3, 23, 2)
(2, '제라드호잉', 0.28800000000000003, 58, 15)
(3, '이성열', 0.255, 62, 18)
(4, '김태균', 0.308, 43, 5)
(5, '정은원', 0.281, 47, 5)
(6, '변우혁', 0.226, 2, 1)
(7, '노시환', 0.19, 12, 1)
(8, '유장혁', 0.174, 3, 1)
(9, '강경학', 0.213, 6, 0)


In [95]:
sql = "SELECT Eagles.back_no, Eagles.name, Eagles.position, Stats.average, Stats.rbi, \
Stats.homerun FROM Eagles JOIN Stats ON Eagles.name like Stats.player;"

cur = conn.cursor()
cur.execute(sql)
for row in cur:
    print(row)

(13, '최재훈', '포수', 0.3, 23, 2)
(30, '제라드호잉', '외야수', 0.28800000000000003, 58, 15)
(50, '이성열', '내야수', 0.255, 62, 18)
(52, '김태균', '내야수', 0.308, 43, 5)
(43, '정은원', '내야수', 0.281, 47, 5)
(27, '변우혁', '내야수', 0.226, 2, 1)
(37, '노시환', '내야수', 0.19, 12, 1)
(66, '유장혁', '외야수', 0.174, 3, 1)


In [97]:
import pandas as pd

In [99]:
sql = "SELECT e.back_no, e.name, e.position, s.average, s.rbi, s.homerun \
FROM Eagles AS e JOIN Stats AS s ON e.name like s.player;"
cur = conn.cursor()
cur.execute(sql)
rows = cur.fetchall();
columnName = ['등번호', '선수명', '포지션', '타율', '타점', '홈런']
eagles_df = pd.DataFrame(columns = columnName) # 빈 데이터 프레임 만들기
for row in rows:
    eagles_df = eagles_df.append(pd.DataFrame([list(row)], 
                                             columns = columnName), ignore_index = True)
eagles_df

등번호    선수명  포지션     타율  타점  홈런
0  13    최재훈   포수  0.300  23   2
1  30  제라드호잉  외야수  0.288  58  15
2  50    이성열  내야수  0.255  62  18
3  52    김태균  내야수  0.308  43   5
4  43    정은원  내야수  0.281  47   5
5  27    변우혁  내야수  0.226   2   1
6  37    노시환  내야수  0.190  12   1
7  66    유장혁  외야수  0.174   3   1

In [1]:
conn.close()

NameError: name 'conn' is not defined